File: webapp/lib/deck.rb
# frozen_string_literal: true

# A drawable, discardable, shuffleable pile of cards. Generic over card type.
# Knows nothing about game rules — only how to draw and discard.
class Deck
  # @param cards [Array] the cards to seed the draw pile with
  # @param rng [Random] optional rng for deterministic shuffles (testing)
  def initialize(cards = [], rng: Random.new)
    @draw_pile = cards.dup
    @discard_pile = []
    @rng = rng
  end

  def shuffle!
    @draw_pile.shuffle!(random: @rng)
    self
  end

  # True only when both piles are empty (nothing left to draw or reshuffle).
  def empty?
    @draw_pile.empty? && @discard_pile.empty?
  end

  def size
    @draw_pile.size
  end

  # Draw a single card. If the draw pile is empty, reshuffle the discard pile
  # back into it first. Returns nil only when both piles are empty.
  def draw
    reshuffle_discards if @draw_pile.empty?
    @draw_pile.shift
  end

  # Draw up to n cards (fewer if the pile runs out).
  def draw_many(n)
    Array.new(n) { draw }.compact
  end

  def discard(card)
    @discard_pile << card if card
    self
  end

  # Take a specific card back out of the discard pile (e.g. to undo a discard).
  # Returns the card if it was there, else nil.
  def reclaim(card)
    @discard_pile.delete(card) ? card : nil
  end

  private

  # Move the discard pile into the draw pile and shuffle it.
  def reshuffle_discards
    return if @discard_pile.empty?

    @draw_pile = @discard_pile.shuffle(random: @rng)
    @discard_pile = []
  end
end

File: android/app/src/main/java/com/dungeonboss/game/Deck.kt
package com.dungeonboss.game

import kotlin.random.Random

/**
 * A drawable, discardable, shuffleable pile of cards. Generic over card type.
 * Knows nothing about game rules — only how to draw and discard. Reshuffles its
 * discard pile back in when drawn from empty.
 */
class Deck<T>(cards: List<T> = emptyList(), private val rng: Random = Random.Default) {
    private var drawPile: MutableList<T> = cards.toMutableList()
    private var discardPile: MutableList<T> = mutableListOf()

    fun shuffle(): Deck<T> {
        drawPile.shuffle(rng)
        return this
    }

    /** True only when both piles are empty (nothing left to draw or reshuffle). */
    fun isEmpty(): Boolean = drawPile.isEmpty() && discardPile.isEmpty()

    fun size(): Int = drawPile.size

    /**
     * Draw a single card. If the draw pile is empty, reshuffle the discard pile
     * back into it first. Returns null only when both piles are empty.
     */
    fun draw(): T? {
        if (drawPile.isEmpty()) reshuffleDiscards()
        return if (drawPile.isEmpty()) null else drawPile.removeAt(0)
    }

    /** Draw up to n cards (fewer if the pile runs out). */
    fun drawMany(n: Int): List<T> = (0 until n).mapNotNull { draw() }

    fun discard(card: T?): Deck<T> {
        if (card != null) discardPile.add(card)
        return this
    }

    /**
     * Take a specific card back out of the discard pile (e.g. to undo a discard).
     * Returns the card if it was there (by identity), else null.
     */
    fun reclaim(card: T): T? {
        val index = discardPile.indexOfFirst { it === card }
        return if (index >= 0) { discardPile.removeAt(index); card } else null
    }

    private fun reshuffleDiscards() {
        if (discardPile.isEmpty()) return
        drawPile = discardPile.shuffled(rng).toMutableList()
        discardPile = mutableListOf()
    }
}
